spock-core/src/main/java/org/spockframework/mock/runtime/MockMakerRegistry.java

```diff
@@ -0,0 +1,154 @@
+package org.spockframework.mock.runtime;
+
+import org.spockframework.mock.CannotCreateMockException;
+import org.spockframework.mock.IMockObject;
+import org.spockframework.mock.ISpockMockObject;
+import org.spockframework.util.InternalSpockError;
+import org.spockframework.util.Nullable;
+import org.spockframework.util.ThreadSafe;
+import spock.mock.MockMakerId;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+
+import static org.spockframework.mock.runtime.IMockMaker.IMockCreationSettings;
+import static org.spockframework.mock.runtime.IMockMaker.ITypeMockability;
+
+@ThreadSafe
+public final class MockMakerRegistry {
+
+  private static final String NL = "\n";
+
+  public static MockMakerRegistry INSTANCE = createFromServiceLoader();
+
+  private final Map<String, IMockMaker> makerMap = new HashMap<>();
+
+  private final List<IMockMaker> makerList;
+
+  private static MockMakerRegistry createFromServiceLoader() {
+    ServiceLoader<IMockMaker> mockMakers = ServiceLoader.load(IMockMaker.class);
+    String preferredMockMakerId = System.getProperty(IMockMaker.MOCK_MAKER_PROPERTY);
+    return new MockMakerRegistry(mockMakers, preferredMockMakerId);
+  }
+
+  MockMakerRegistry(Iterable<IMockMaker> mockMakers, @Nullable String preferredMockMakerIdParam) {
+    List<IMockMaker> makerListLoc = new ArrayList<>();
+    mockMakers.forEach(m -> {
+      makerListLoc.add(m);
+      String id = m.getId();
+      if (id == null || !id.endsWith("-mock-maker")) {
```

Comment:
I'm not so sure about the `-mock-maker` suffix requirement. What do you hope to prevent with it? The most common error case would be a typo in the test code, but that will fail just as well without the suffix.

Reply:
The reason behind this was for somewhat consistent error messages like:

```
Cannot create mock for class java.lang.StringBuilder.
dynamic-proxy-mock-maker: Type is not an interface.
byte-buddy-mock-maker: Cannot mock final classes.
cglib-mock-maker: Cannot mock final classes.
```
The suffix ensures that an error message containing that ID can be linked to the `IMockMaker`.
If e.g. another part will be involved it is more distingishable, e.g.:

```
Cannot create mock for class java.lang.StringBuilder.
dynamic-proxy-mock-maker: Type is not an interface.
objenesis-instatiator: Library not found.
```

Reply:
I have changed it to `MockMakerId`.